### PR TITLE
Clean `kubernetes.client`, add controller-runtime

### PR DESCRIPTION
**Pull request by adracus**
**What this PR does / why we need it**:
Cleanup of client code in general & as first step to improve the
whole cleanup code we currently have.

Steps done so far:
* rename `kubernetes.Client` to `kubernetes.Interface`
* rename clientset methods and change their type to the respective
interface
* move `kubernetes.base` to `kubernetes`
* remove unneeded `v*` packages (version changes have to be handled explicitly in code now [though it is currently not necessary])
* rename `kubernetes.base.Client` to `kubernetes.Clientset`
* delete unneeded methods from `kubernetes.Interface`
* move out `kubernetes.Interface.Apply` to `ApplierInterface` / `Applier` and create tests
* update current methods to use new coding / delete old methods (e.g. machine client REST API usage)

```improvement operator
NONE
```

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -38,6 +38,10 @@ ignored = [
   name = "k8s.io/kube-aggregator"
   version = "kubernetes-1.12.1"
 
+[[override]]
+  name = "sigs.k8s.io/controller-runtime"
+  revision = "a4051585e748b437e799d597f9adba1d777cb12a"
```
Comment (mvladev) [inline Gopkg.toml:44]: can you comment on which version this referencing to?

**Comment by rfranzke** [comment]
(on hold until 0.14.0 is released)

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go [hunk 2]
```diff
@@ -250,33 +251,35 @@ func NewGardener(config *componentconfig.ControllerManagerConfiguration) (*Garde
 		)
 	)
 
-	k8sGardenClient, err := kubernetes.NewClientFromFile(kubeconfig, &config.ClientConnection)
+	k8sGardenClient, err := kubernetes.NewClientFromFile(kubeconfig, &config.ClientConnection, client.Options{
+		Scheme: kubernetes.GardenScheme,
```
Comment (zanetworker) [inline cmd/gardener-controller-manager/app/gardener_controller_manager.go:255]: Do we need to explicitly specify the scheme here?  it is defaulted to `var Scheme = runtime.NewScheme()` anyway when not set.

**Review by zanetworker — COMMENTED** [review]
Please recheck vendored packages as some seem to updated (e.g., miekg/dns).

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go (hunk 2 above)
Comment (adracus) [reply cmd/gardener-controller-manager/app/gardener_controller_manager.go:255]: `GardenScheme` is not only `runtime.NewScheme` but also gets all the schemes from `corescheme` and `gardenscheme` (see `init()` function in `kubernetes/types.go`)

**Review by rfranzke — APPROVED** [review]
/lgtm, nice! Thank you

**Pull request merged**